test/parallel_api/ranges/std_ranges_test.h

```diff
@@ -0,0 +1,455 @@
+
+            //check result
+            if constexpr(RetTypeCheck)
+                static_assert(std::is_same_v<decltype(res), decltype(checker(tr(A), args...))>, "Wrong return type");
+
+            auto bres = ret_in_val(expected_res, expected_view.begin()) == ret_in_val(res, tr(A).begin());
+            EXPECT_TRUE(bres, (std::string("wrong return value from algo with ranges: ") + typeid(Algo).name() + 
+                typeid(decltype(tr(std::declval<Container&>()()))).name()).c_str());
+        }
+
+        //check result
+        EXPECT_EQ_N(expected, data, max_n, (std::string("wrong effect algo with ranges: ")
+            + typeid(Algo).name() + typeid(decltype(tr(std::declval<Container&>()()))).name()).c_str());
+    }
+    template<typename Policy, typename Algo, typename Checker, typename Transform>
+    std::enable_if_t<!std::is_same_v<Policy, std::true_type> && Ranges == data_in_out>
+    operator()(Policy&& exec, Algo algo, Checker checker, Transform tr, auto... args)
+    {
+        constexpr int max_n = 10;
+        DataType data_in[max_n] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+        DataType data_out[max_n] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+        DataType expected[max_n] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+
+        auto src_view = tr(std::ranges::subrange(data_in, data_in + max_n));
+        auto expected_res = checker(src_view, expected, args...);
+        {
+            Container cont_in(exec, data_in, max_n);
+            Container cont_out(exec, data_out, max_n);
+
+            typename Container::type& A = cont_in();
+            typename Container::type& B = cont_out();
+
+            auto res = algo(exec, tr(A), B, args...);
+
+            //check result
+            if constexpr(RetTypeCheck)
+                static_assert(std::is_same_v<decltype(res), decltype(checker(tr(A), B, args...))>, "Wrong return type");
+
+            auto bres_in = ret_in_val(expected_res, src_view.begin()) == ret_in_val(res, tr(A).begin());
+            EXPECT_TRUE(bres_in, (std::string("wrong return value from algo with input range: ") + typeid(Algo).name()).c_str());
+
+            auto bres_out = ret_out_val(expected_res, expected) == ret_out_val(res, B.begin());
+            EXPECT_TRUE(bres_out, (std::string("wrong return value from algo with output range: ") + typeid(Algo).name()).c_str());
+        }
+
+        //check result
+        EXPECT_EQ_N(expected, data_out, max_n, (std::string("wrong effect algo with ranges: ") + typeid(Algo).name()).c_str());
+    }
+
+    template<typename Policy, typename Algo, typename Checker, typename Transform>
+    std::enable_if_t<!std::is_same_v<Policy, std::true_type> && Ranges == data_in_in>
+    operator()(Policy&& exec, Algo algo, Checker checker, Transform tr, auto... args)
+    {
+        constexpr int max_n = 10;
+        DataType data_in1[max_n] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+        DataType data_in2[max_n] = {0, 0, 2, 3, 4, 5, 0, 0, 0, 0};
```

Comment:
It would be perfect to test cases when the result is found or not found: the newly added layer has some related logic in `__pattern_search` while deciding which iterator to return.

Reply:
A case "not find" was added.